docs/knowledgebase/advanced/try-runtime.md

```diff
@@ -0,0 +1,185 @@
+
+## Motivation
+
+The initial motivation for `try-runtime` came from the need to extend the capabilities of [`TestExternalities`][testextern-rustdocs] and [`BasicExternalities`][basicextern-rustdocs] 
+in such a way that it would be possible to use externalities to populate its memory with real node data. Since externalities provide the capability of storing an in-memory
+hash-map, extending these capabilities is achieved by simply using the key-value pairs from a node's RPCs.
+
+## How it works
+The `try-runtime` tool has its own implementation of externalities called [`remote_externalities`][remoteextern-rustdocs], with its hash-map containing:
+
+- a **Key**: the `Hash` of `hash(pallet_name) + hash(storage)`; and
+- a **Value**: some SCALE encoded `Vec<u8>` Rust value. 
+
+The diagram below illustrates the way externalities sits outside a compiled runtime as a means to capture 
+the storage of that runtime. 
+
+Storage externalities            |  Testing with externalities
+:-------------------------:|:-------------------------:
+![image](./../assets/advanced/try-runtime-ext-1.png)  |  ![image](./../assets/advanced/try-runtime-ext-2.png)
+
+With `remote_externalities`, developers can capture some chain state and run tests on it. By contrast, `TestExternalities` only allows developers to 
+debug old state and `BasicExternalities` is designed only to be used in genesis. 
+
+In order to query state, `try-runtime` makes use of Substrate Core's [`StateApi`][stateapi-rustdocs]. In particular:
+- [`storage`][stateapi-storage-rustdocs]: A method which returns a storage entry at a specific block’s state.
+- [`storage_key_paged`][stateapi-storage-keys-paged-rustdocs]: A method which returns the keys with prefix with pagination support.
+
+// TODO: Add about frame::Executor
+
+## Usage
+
+The most common use case for `try-runtime` is with storage migrations and runtime upgrades.
+
+### Calling into hooks from `OnRuntimeUpgrade`
+There's two ways of calling into `OnRuntimeUpgrade` hooks:
+- **From inside a runtime**. For example:
+    ```rust
+    struct Custom;
+    impl OnRuntimeUpgrade for Custom {
+        fn on_runtime_upgrade() -> Weight { }
+    }
+    ```
+
+- **From a pallet**. For example:
+    ```rust
+    #[pallet::hooks]
+    fn on_runtime_upgrade() -> Weight { }
+    ```
+
+To call into these hooks with `try-runtime`, a `feature` tag is used. This requires specifying a `features` 
+dependency in your projects' relevant `Cargo.toml` files. 
+
+> Refer to this how-to guide on integrating `try-runtime` to your project. (TODO)
+
+With this feature tag, `try-runtime` can call into `OnRuntimeUpgrade` hooks using the 
+`#[cfg(feature = "try-runtime")]` macro. This macro must be 
+declared before each function being tested in order for `try-runtime` to capture the relevant state 
+that its user specified. For example, in a storage migration, `try-runtime` can allow developers to 
```

Comment:
This section is a bit superfolous. I am not sure what 

> to capture the relevant state that its user specified

 means. 

I think what you are trying to say is: 

- there are two ways to implement on-runtime upgrade. These are methods are `OnRuntimeUpgrade` trait.
- if the feature is enabled, you can implement two additional methods as well.
- if you do so, these methods are called when you use `try-runtime on-runtime-upgrade`.